File: MOBILE APP/exhibition_companion/lib/state/app_state.dart
import 'dart:async';
import 'dart:io' show Platform;

// 'Category' here is Flutter's dartdoc annotation, which collides with the
// exhibition's own Category model. We never use the annotation.
import 'package:flutter/foundation.dart' hide Category;
import 'package:shared_preferences/shared_preferences.dart';

import '../api/api_client.dart';
import '../api/models.dart';
import '../config.dart';

enum AuthStage {
  /// Reading the saved token from the device.
  starting,

  /// Nobody signed in.
  signedOut,

  /// Signed in and the exhibition is loaded.
  signedIn,
}

/// The one piece of shared state in the app: who is signed in, which exhibition
/// system we are talking to, and the reference data every screen filters by.
///
/// Screens own their own lists — a search result belongs to the search screen —
/// so this stays small. What lives here is only what more than one screen needs
/// and what must survive navigation: the session, the category tree and hall
/// list behind the filters, and the count of catalogues collected today, which
/// the scanner writes and the tab badge reads.
class AppState extends ChangeNotifier {
  AppState({ApiClient? client}) : _client = client ?? ApiClient(baseUrl: AppConfig.defaultBaseUrl) {
    _client.onUnauthorised = _onServerRejectedToken;
  }

  ApiClient _client;
  ApiClient get api => _client;

  AuthStage _stage = AuthStage.starting;
  AuthStage get stage => _stage;

  Visitor? _visitor;
  Visitor? get visitor => _visitor;

  Exhibition? _exhibition;
  Exhibition? get exhibition => _exhibition;

  String _baseUrl = AppConfig.defaultBaseUrl;
  String get baseUrl => _baseUrl;

  String? _lastEmail;
  String? get lastEmail => _lastEmail;

  /// Set when the token was rejected mid-session, so the sign-in screen can
  /// explain why the visitor is suddenly back there.
  String? _signedOutBecause;
  String? get signedOutBecause => _signedOutBecause;

  int _cataloguesToday = 0;
  int get cataloguesToday => _cataloguesToday;

  List<Hall> get halls => _exhibition?.halls ?? const [];
  List<Category> get categories => _exhibition?.categories ?? const [];
  List<String> get countries => _exhibition?.countries ?? const [];
  List<DateTime> get programmeDates => _exhibition?.programmeDates ?? const [];

  /// Sub-categories of one category, for the second filter dropdown.
  List<Category> subCategoriesOf(int? categoryId) {
    if (categoryId == null) return const [];
    for (final category in categories) {
      if (category.id == categoryId) return category.children;
    }
    return const [];
  }

  // --- start-up ------------------------------------------------------------

  /// Restores the saved session, if the token is still good.
  Future<void> restore() async {
    final prefs = await SharedPreferences.getInstance();

    _baseUrl = prefs.getString(AppConfig.baseUrlKey) ?? AppConfig.defaultBaseUrl;
    _lastEmail = prefs.getString(AppConfig.emailKey);
    _rebuildClient(_baseUrl);

    final token = prefs.getString(AppConfig.tokenKey);
    if (token == null || token.isEmpty) {
      _stage = AuthStage.signedOut;
      notifyListeners();
      return;
    }

    _client.token = token;

    try {
      _visitor = await _client.me();
      await _loadExhibition();
      _stage = AuthStage.signedIn;
    } on ApiException catch (e) {
      // A rejected token means the session is genuinely over. A network
      // failure does not — the venue's wifi being down at 08:55 must not sign
      // every visitor out, so the token is kept and the app retries.
      if (e.isUnauthorised) {
        await _clearToken();
        _stage = AuthStage.signedOut;
      } else {
        _stage = AuthStage.signedOut;
        _signedOutBecause = e.message;
      }
    }

    notifyListeners();
  }

  /// Point the app at a different exhibition system. Signs out, because a token
  /// from one server means nothing to another.
  Future<void> setBaseUrl(String url) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(AppConfig.baseUrlKey, url);
    await prefs.remove(AppConfig.tokenKey);

    _baseUrl = url;
    _rebuildClient(url);
    _visitor = null;
    _exhibition = null;
    _stage = AuthStage.signedOut;
    notifyListeners();
  }

  void _rebuildClient(String url) {
    _client.close();
    _client = ApiClient(baseUrl: url)..onUnauthorised = _onServerRejectedToken;
  }

  // --- signing in ----------------------------------------------------------

  Future<LoginCodeRequest> requestCode(String email) async {
    final result = await _client.requestLoginCode(email.trim());

    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(AppConfig.emailKey, email.trim());
    _lastEmail = email.trim();

    return result;
  }

  Future<void> verifyCode(String email, String code) async {
    final result = await _client.verifyLoginCode(
      email: email.trim(),
      code: code.trim(),
      platform: _platformName,
      deviceName: _deviceLabel,
      appVersion: AppConfig.appVersion,
    );

    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(AppConfig.tokenKey, result.token);

    _visitor = result.visitor;
    _signedOutBecause = null;

    await _loadExhibition();

    _stage = AuthStage.signedIn;
    notifyListeners();
  }

  Future<void> signOut() async {
    await _client.logout();
    await _clearToken();

    _visitor = null;
    _exhibition = null;
    _cataloguesToday = 0;
    _signedOutBecause = null;
    _stage = AuthStage.signedOut;
    notifyListeners();
  }

  Future<void> _clearToken() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(AppConfig.tokenKey);
    _client.token = null;
  }

  /// The server told us the token is no longer valid, from wherever in the app
  /// that happened. Tear the session down once rather than in every screen.
  void _onServerRejectedToken() {
    if (_stage != AuthStage.signedIn) return;

    _stage = AuthStage.signedOut;
    _visitor = null;
    _exhibition = null;
    _signedOutBecause =
        'Your session has ended. Sign in with your registered email address to continue.';

    unawaited(_clearToken());
    notifyListeners();
  }

  // --- shared data ---------------------------------------------------------

  Future<void> _loadExhibition() async {
    _exhibition = await _client.exhibition();
    await refreshCatalogueCount();
  }

  /// Reloads the reference data. Cheap, and the programme genuinely changes
  /// during a show.
  Future<void> refresh() async {
    await _loadExhibition();
    notifyListeners();
  }

  Future<void> refreshCatalogueCount() async {
    try {
      _cataloguesToday = (await _client.myCatalogues()).length;
      notifyListeners();
    } on ApiException {
      // The badge on a tab is not worth interrupting anyone for.
    }
  }

  /// Called by the scanner, which already knows the new count.
  void setCatalogueCount(int count) {
    _cataloguesToday = count;
    notifyListeners();
  }

  Future<void> updateConsent({bool? email, bool? tracking}) async {
    _visitor = await _client.updateConsent(email: email, tracking: tracking);
    notifyListeners();
  }

  static String get _platformName {
    if (kIsWeb) return 'web';
    if (Platform.isAndroid) return 'android';
    if (Platform.isIOS) return 'ios';
    return Platform.operatingSystem;
  }

  static String get _deviceLabel {
    if (kIsWeb) return 'Browser';
    return Platform.operatingSystemVersion.split('(').first.trim();
  }

  @override
  void dispose() {
    _client.close();
    super.dispose();
  }
}

File: src/Exb.Mobile.Shared/Services/AppState.cs
using Exb.Mobile.Shared.Models;

namespace Exb.Mobile.Shared.Services;

public enum AuthStage { Starting, SignedOut, SignedIn }

/// <summary>
/// App-wide session state — the C# equivalent of the Flutter app's
/// ChangeNotifier-based AppState/AppScope: a scoped DI service the whole
/// component tree observes via <see cref="Changed"/>. See the porting notes
/// for why the 401 vs. other-error distinction in <see cref="RestoreAsync"/>
/// matters (a dead venue wifi at opening time must not silently and
/// permanently sign everyone out).
/// </summary>
public sealed class AppState
{
    private const string TokenKey = "exb.token";
    private const string BaseUrlKey = "exb.baseUrl";
    private const string EmailKey = "exb.email";

    private readonly IPlatformServices _platform;
    private readonly Func<string, ApiClient> _clientFactory;

    public AppState(IPlatformServices platform, Func<string, ApiClient> clientFactory, string defaultBaseUrl)
    {
        _platform = platform;
        _clientFactory = clientFactory;
        BaseUrl = defaultBaseUrl;
        Api = _clientFactory(BaseUrl);
        Api.Unauthorised += OnServerRejectedToken;
    }

    public event Action? Changed;
    private void Notify() => Changed?.Invoke();

    public ApiClient Api { get; private set; }
    public AuthStage Stage { get; private set; } = AuthStage.Starting;
    public Visitor? Visitor { get; private set; }
    public Exhibition? Exhibition { get; private set; }
    public string BaseUrl { get; private set; }
    public string? LastEmail { get; private set; }
    public string? SignedOutBecause { get; private set; }
    public int CataloguesToday { get; private set; }

    public IReadOnlyList<Hall> Halls => Exhibition?.Halls ?? [];
    public IReadOnlyList<Category> Categories => Exhibition?.Categories ?? [];
    public IReadOnlyList<string> Countries => Exhibition?.Countries ?? [];
    public IReadOnlyList<DateOnly> ProgrammeDates => Exhibition?.ProgrammeDates ?? [];

    public IReadOnlyList<Category> SubCategoriesOf(int? categoryId)
    {
        if (categoryId is null) return [];
        return Categories.FirstOrDefault(c => c.Id == categoryId)?.Children ?? [];
    }

    /// <summary>
    /// Restores the saved session at launch. Whatever happens, this must end
    /// with the app on either the login screen or the home shell: the splash
    /// it runs behind has no controls at all, so any escape from here strands
    /// the visitor on a spinner with nothing to tap.
    /// </summary>
    public async Task RestoreAsync()
    {
        try
        {
            BaseUrl = await _platform.GetPreference(BaseUrlKey) ?? BaseUrl;
            LastEmail = await _platform.GetPreference(EmailKey);
            RebuildClient();

            string? token = await _platform.GetSecureToken();
            if (string.IsNullOrEmpty(token))
            {
                Stage = AuthStage.SignedOut;
                Notify();
                return;
            }

            Api.Token = token;
            try
            {
                Visitor = await Api.Me();
                await LoadExhibitionAsync();
                Stage = AuthStage.SignedIn;
            }
            catch (ApiException ex)
            {
                if (ex.IsUnauthorised)
                {
                    await _platform.SetSecureToken(null);
                }
                else
                {
                    // Keep the token — a network hiccup at launch should not
                    // permanently sign the visitor out; they can retry.
                    SignedOutBecause = ex.Message;
                }
                Stage = AuthStage.SignedOut;
            }
        }
        catch (Exception ex)
        {
            // Anything the API client did not turn into an ApiException — a
            // malformed stored address, unreadable secure storage, a response
            // that would not deserialise — used to escape this method and
            // leave the splash spinner on screen for good. Land on the login
            // screen instead and say what went wrong.
            SignedOutBecause = $"The app could not start its last session ({ex.GetType().Name}). Sign in again, or check the address below.";
            Stage = AuthStage.SignedOut;
        }
        Notify();
    }

    public async Task SetBaseUrlAsync(string url)
    {
        BaseUrl = url;
        await _platform.SetPreference(BaseUrlKey, url);
        await _platform.SetSecureToken(null);
        RebuildClient();
        Visitor = null;
        Exhibition = null;
        Stage = AuthStage.SignedOut;
        Notify();
    }

    private void RebuildClient()
    {
        Api.Unauthorised -= OnServerRejectedToken;
        Api = _clientFactory(BaseUrl);
        Api.Unauthorised += OnServerRejectedToken;
    }

    public async Task<LoginCodeRequest> RequestCodeAsync(string email)
    {
        string trimmed = email.Trim();
        var result = await Api.RequestLoginCode(trimmed);
        LastEmail = trimmed;
        await _platform.SetPreference(EmailKey, trimmed);
        return result;
    }

    public async Task VerifyCodeAsync(string email, string code)
    {
        var (token, visitor) = await Api.VerifyLoginCode(
            email, code, _platform.PlatformName, _platform.DeviceLabel, AppConfig.AppVersion);

        await _platform.SetSecureToken(token);
        Visitor = visitor;
        SignedOutBecause = null;
        await LoadExhibitionAsync();
        Stage = AuthStage.SignedIn;
        Notify();
    }

    public async Task SignOutAsync()
    {
        // Sign out locally first and tell the server afterwards. Waiting on
        // the network here means that when the venue wifi is slow — exactly
        // when somebody is most likely to be giving up and signing out — the
        // button appears to do nothing for the length of the HTTP timeout.
        // Revoking the token server-side is worth attempting but is not worth
        // making the visitor wait for, and the local token is gone either way.
        var client = Api;

        await _platform.SetSecureToken(null);
        Visitor = null;
        Exhibition = null;
        CataloguesToday = 0;
        SignedOutBecause = null;
        Stage = AuthStage.SignedOut;
        Notify();

        _ = Task.Run(async () =>
        {
            try { await client.Logout(); } catch { /* best-effort revocation */ }
        });
    }

    private void OnServerRejectedToken()
    {
        if (Stage != AuthStage.SignedIn) return;
        Stage = AuthStage.SignedOut;
        Visitor = null;
        Exhibition = null;
        SignedOutBecause = "Your session has ended. Sign in with your registered email address to continue.";
        _ = _platform.SetSecureToken(null);
        Notify();
    }

    private async Task LoadExhibitionAsync()
    {
        Exhibition = await Api.GetExhibition();
        await RefreshCatalogueCountAsync();
    }

    public async Task RefreshAsync()
    {
        await LoadExhibitionAsync();
        Notify();
    }

    public async Task RefreshCatalogueCountAsync()
    {
        try { CataloguesToday = (await Api.MyCatalogues()).Count; }
        catch (ApiException) { /* a badge count failing shouldn't interrupt anyone */ }
    }

    public void SetCatalogueCount(int count)
    {
        CataloguesToday = count;
        Notify();
    }

    public async Task UpdateConsentAsync(bool? email = null, bool? tracking = null)
    {
        Visitor = await Api.UpdateConsent(email, tracking);
        Notify();
    }
}

public static class AppConfig
{
    public const string AppVersion = "1.0.0";
}
